src/aws/osml/model_runner/database/ddb_helper.py

```diff
@@ -104,6 +105,40 @@ def put_ddb_item(self, ddb_item: DDBItem, condition_expression: str = None) -> D
 
         return response
 
+    def batch_write_items(self, ddb_items: List[DDBItem], max_retries: int = 40, max_delay: float = 8):
+        """
+        Write multiple DynamoDB items in a batch with retry logic for unprocessed items.
+        :param ddb_items: List[DDBItem] = List of items that we want to write.
+        :param max_retries: int = Maximum number of retries for unprocessed items.
+        :param max_delay: Maximum delay in seconds between retries.
+        :return: None
+        """
+
+        def _batch_write(items: Dict[str, Any], retries=0):
+            response = self.client.batch_write_item(RequestItems=items)
+            unprocessed_items = response.get("UnprocessedItems", {})
+            # If we failed to process some items, try again
+            if unprocessed_items and retries < max_retries:
+                # Calculate jitter for retries and wait
+                time.sleep(random.uniform(1, max_delay))
```

Comment:
Why not use the exponential backoff that is recommended with the DDB batch writer?

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html#Programming.Errors.BatchOperations

https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/

Reply:
Because the actual batch_write() operation is already using an logical backoff like that and this is to try the unprocessed items associated with this specific batch, which we don't want to get hung up in the exponential backoff and fall behind.